csrc/id_model/id_model.cpp

```diff
@@ -0,0 +1,292 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <id_model/id_model.h>
+#include <id_model/to_string.h>
+
+#include <device_lower/analysis/trivial_broadcast.h>
+#include <device_lower/lower2device.h>
+#include <device_lower/utils.h>
+#include <disjoint_set.h>
+#include <ir/utils.h>
+#include <root_domain_map.h>
+#include <transform_iter.h>
+
+#include <tuple>
+#include <typeinfo>
+#include <utility>
+
+namespace nvfuser {
+
+namespace {
+
+// Map through loop swizzles, as input/output IterDomains are exact, only the
+// order they're traversed differs.
+void mapThroughLoopSwizzles(ValGraph& graph) {
+  std::vector<Swizzle2D*> all_swizzles;
+
+  for (const auto& expr_set :
+       std::as_const(graph).disjointExprSets().disjointSets()) {
+    auto swizzles_in_expr_set = ir_utils::filterByType<Swizzle2D>(
+        expr_set->vector().begin(), expr_set->vector().end());
+    all_swizzles.insert(
+        all_swizzles.end(),
+        swizzles_in_expr_set.begin(),
+        swizzles_in_expr_set.end());
+  }
+
+  for (auto swizzle : all_swizzles) {
+    if (swizzle->swizzleMode() == SwizzleMode::Loop) {
+      graph.mapVals(swizzle->inX(), swizzle->outX());
+      graph.mapVals(swizzle->inY(), swizzle->outY());
+    }
+  }
+}
+
+} // namespace
+
+IdModel::IdModel(
+    const std::vector<Expr*>& exprs,
+    const std::vector<TensorView*>& additional_tvs,
+    bool allow_self_mapping) {
+  build(exprs, additional_tvs);
+}
+
+IdModel::IdModel(const std::vector<Expr*>& exprs, bool allow_self_mapping)
+    : IdModel(exprs, {}, allow_self_mapping) {}
+
+IdModel::IdModel(Fusion* fusion, bool allow_self_mapping) {
+  std::vector<TensorView*> inputs_and_outputs;
+  {
+    auto inp_tvs = ir_utils::filterByType<TensorView>(fusion->inputs());
+    inputs_and_outputs.insert(
+        inputs_and_outputs.begin(), inp_tvs.begin(), inp_tvs.end());
+  }
+  {
+    auto out_tvs = ir_utils::filterByType<TensorView>(fusion->outputs());
+    inputs_and_outputs.insert(
+        inputs_and_outputs.begin(), out_tvs.begin(), out_tvs.end());
+  }
+
+  build(fusion->exprs(), inputs_and_outputs);
+}
+
+const ValGraph& IdModel::idGraph(IdMappingMode mode) const {
+  auto graph_it = id_graphs_.find(mode);
+  NVF_ERROR(graph_it != id_graphs_.end());
+  return graph_it->second;
+}
+
+ValGraph& IdModel::idGraph(IdMappingMode mode) {
+  auto graph_it = id_graphs_.find(mode);
+  NVF_ERROR(graph_it != id_graphs_.end());
+  return graph_it->second;
+}
+
+Expr* IdModel::idUse(IterDomain* id) const {
+  auto use_it = id_uses_.find(id);
+  if (use_it == id_uses_.end()) {
+    return nullptr;
+  }
+  return use_it->second.front();
```

Comment:
Here, it's actually fine to have multiple definitions or uses. This interface is to just grab whichever expression that is grouped as a use or a def, which can be different from the actual use or def expressions as long as they are grouped together. For example,

```
// tv0: [i0]
auto tv0 = makeSymbolicTensor(1);

// tv1: [i1]
auto tv1 = set(tv0);

// tv2: [i2]
auto tv2 = set(tv0);

// tv1: [i1_1, i1_2]
tv1->split(0, 4);

// tv2: [i2_1, i2_2]
tv2->split(0, 4);
```

In this case, `idDef(i1_1)` may return the split of `tv2` since it's grouped with the split of `tv1`.

I'm not sure if these interfaces are actually used. I'll revisit and clean up if necessary. #1235

Reply:
Thanks for clarifying!